Add standard library DUI  content to the release note (#114)
ballerina-platform/ballerina-release: release-notes/swan-lake-alpha3-release-note.md

#### Standard Library

##### Log Package Updates

###### Introduce additional log levels and log functions

> There are 4 log levels: `DEBUG`, `ERROR`, `INFO`,and `WARN` and their respective functions: `printDebug`, `printError`, `printInfo`, and `printWarn`. To set the global log level, place the entry given below in the `Config.toml` file:

Reviewer: ```suggestion
There are 4 log levels: `DEBUG`, `ERROR`, `INFO`, and `WARN` and their respective functions: `printDebug`, `printError`, `printInfo`, and `printWarn`. To set the global log level, place the entry given below in the `Config.toml` file:
```

```
[log] 
level = "[LOG_LEVEL]"
```

Each module can also be assigned its own log level. To assign a log level to a module, provide the following entry in the `Config.toml` file:
```
[[log.modules]] 
name = "[MODULE_NAME]" 
level = "[LOG_LEVEL]"
```

#### OS Package Updates

- Removed the `exec` function.

#### Task Package Updates

> The module has been revamped by removing `Scheduler` and `Listener` classes and introducing the following functions to schedule and manage the job either one-time or periodically.

Reviewer 2: ```suggestion
The module has been revamped by removing the `Scheduler` and `Listener` classes and introducing the following functions to schedule and manage the job either one-time or periodically.
```

> - Configure the scheduler worker pool with worker count and max waiting time.

Reviewer 2: ```suggestion
- Configure the scheduler worker pool with the worker count and max waiting time.
```

```ballerina
task:Error? output = task:configureWorkerPool(6, 7000);
```

> - Schedule the Ballerina job at specified time.

Reviewer 2: ```suggestion
- Schedule the Ballerina job at a specified time.
```

```ballerina
class MyJob {

>    *task:Job;

Reviewer: Shall we add the import for task?

Reviewer 3: Added.

   public function execute() {
       // logic goes here
   }
}

ZoneOffset zoneOffset = {hours: 5, minutes: 30};

> Civil time = {

Reviewer: Do these come from the time package?

Reviewer 4: Yes, fixed in here https://github.com/ballerina-platform/ballerina-release/pull/114#discussion_r595718951

> ZoneOffset zoneOffset = {hours: 5, minutes: 30};
> Civil time = {

Reviewer 4: ```suggestion
time:ZoneOffset zoneOffset = {hours: 5, minutes: 30};
time:Civil time = {
```

    year: 2021,
    month: 4,
    day: 12,
    hour: 23,
    minute: 20,
    second: 50.52,
    timeAbbrev: "Asia/Colombo",
    utcOffset: zoneOffset
};
task:Error|task:JobId id = task:scheduleOneTimeJob(new MyJob(), time);
```

- Schedule the recurring Ballerina job according to the given duration.
```ballerina
class MyJob {

>    *task:Job;

Reviewer: Shall we add the import for task?

Reviewer 3: Added.

   public function execute() {
       // logic goes here
   }
}
task:Error|task:JobId id = task:scheduleJobRecurByFrequency(new MyJob(), 1);
```

> - Unschedule the particular `job`.

Reviewer: ```suggestion
- Unschedule a job.
```

```ballerina
task:Error? result = unscheduleJob(id);
```

> - Pauses all the jobs.

Reviewer: This is not consistent with the previous one, right?

We should either change "Unschedule" to "Unschedules" or change this (and the following) to "Pauses".

Reviewer 3: Changed.

```ballerina
task:Error? result = pauseAllJobs();
```

- Resumes all the jobs.
```ballerina
task:Error? result = resumeAllJobs();
```

- Pauses the particular job.
```ballerina
task:Error? result = pauseJob(id);
```

- Resumes the particular job.
```ballerina
task:Error? result = resumeJob(id);
```

- Gets all the running jobs.
```ballerina
task:Error? result = getRunningJobs();
```

#### Time Package Updates

Revamped the entire time package as follows:

> - Introduced `time:Utc` to represent the UTC timestamp.

Reviewer: Shall we say
```suggestion
- Introduced the `time:Utc` record to represent the UTC timestamp.
```
or class if it is a class. Applicable to Civil too.

Reviewer 4: Fixed in here https://github.com/ballerina-platform/ballerina-release/pull/114/commits/0ea5a4953d84be5802f7cbf65e6a119655d8c179

- Introduced `time:Civil` to represent the localized time.
- Added necessary APIs to do time generation, manipulations, and conversions.

> Necessary migrations steps from the previous version to the current version are listed [here](https://github.com/ballerina-platform/ballerina-standard-library/issues/1079).

Reviewer: ```suggestion
Steps for migration from the previous version to the current version are listed [here](https://github.com/ballerina-platform/ballerina-standard-library/issues/1079).
```

IINM, we usually try to avoid links with "here". @praneesha FYA please.

Reviewer 2: Yes. Please add a hyperlink to a text. E.g.,

"Necessary migrations steps from the previous version to the current version are listed [in this issue]".,

Author: Updated in 0f32666

#### Cache Package Updates

> - Introduced the new configuration as `EvictionPolicy` to set the eviction policy in the `CacheConfig`.

Reviewer: Is this a record?
```suggestion
- Introduced the new configuration as `EvictionPolicy` to set the eviction policy in the `CacheConfig` record.
```

> The `EvictionPolicy` has been introduced with the option `LRU` as the module only supports the LRU eviction policy to evict the cache data when the cache is full.

Reviewer: Is this a record?
```suggestion
The `EvictionPolicy` record has been introduced with the option `LRU` as the module only supports the LRU eviction policy to evict the cache data when the cache is full.
```

> - Removed the `AbstractEvictionPolicy` Object.

Reviewer 2: ```suggestion
- Removed the `AbstractEvictionPolicy` object.
```

This object had the common APIs for the cache eviction functionalities to implement the custom eviction policy. It has been removed by introducing the above configuration.

#### New `xmldata` Package

> A new module is added to convert a natural representation of data in XML into a natural representation of data in JSON and vice-versa

Reviewer 2: ```suggestion
A new module is added to convert a natural representation of data in XML into a natural representation of data in JSON and vice-versa.
```

- Converts a JSON object to an XML representation.
```ballerina
json data = {
    name: "John",
    age: 30
};
xml|Error x = fromJson(data);
```

> - Converts an XML object to its JSON representation.

Reviewer: ```suggestion
- Converts an XML value to its JSON representation.
```

```ballerina
json|Error j = toJson(xml `foo`);
```

#### Remove `jsonutils`, `xmlutils`, `runtime`, and `reflect` Packages

> The `jsonutils`, `xmlutils`, `runtime` and `reflect` packages are removed from Standard Libraries.

Reviewer 2: ```suggestion
The `jsonutils`, `xmlutils`, `runtime`, and `reflect` packages are removed from Standard Libraries.
```

> The XML/JSON conversation APIs in `jsonutils`, `xmltutils` packages are now supported in `xmldata` package.

Reviewer 2: ```suggestion
The XML/JSON conversation APIs in `jsonutils` and `xmltutils` packages are now supported by the `xmldata` package.
```

#### Standard Library

##### Log Package Updates

###### Introduce additional log levels and log functions

There are 4 log levels: `DEBUG`, `ERROR`, `INFO`, and `WARN` and their respective functions: `printDebug`, `printError`, `printInfo`, and `printWarn`. To set the global log level, place the entry given below in the `Config.toml` file:

> ```

Reviewer: ```suggestion
```toml
```

[log] 
level = "[LOG_LEVEL]"
```

Each module can also be assigned its own log level. To assign a log level to a module, provide the following entry in the `Config.toml` file:

> ```

Reviewer: ```suggestion
```toml
```

};
task:Error|task:JobId id = task:scheduleOneTimeJob(new MyJob(), time);
```

- Schedules the recurring Ballerina job according to the given duration.
```ballerina
import ballerina/task;

class MyJob {
   *task:Job;

   public function execute() {
       // logic goes here
   }
}
task:Error|task:JobId id = task:scheduleJobRecurByFrequency(new MyJob(), 1);
```

- Unschedules the particular job.
```ballerina
import ballerina/task;

task:Error? result = task:unscheduleJob(id);
```

- Pauses all the jobs.
```ballerina
import ballerina/task;

task:Error? result = task:pauseAllJobs();
```

- Resumes all the jobs.
```ballerina
import ballerina/task;

task:Error? result = task:resumeAllJobs();
```

- Pauses the particular job.
```ballerina
import ballerina/task;

task:Error? result = task:pauseJob(id);
```

- Resumes the particular job.
```ballerina
import ballerina/task;

task:Error? result = task:resumeJob(id);
```

- Gets all the running jobs.
```ballerina
import ballerina/task;

task:Error? result = task:getRunningJobs();
```

#### Time Package Updates

Revamped the entire time package as follows:

- Introduced the `time:Utc` record to represent the UTC timestamp.

> - Introduced `time:Civil` to represent the localized time.

Reviewer: ```suggestion
- Introduced the `time:Civil` record to represent the localized time.
```

Reviewer 4: Fixed in here https://github.com/ballerina-platform/ballerina-release/pull/114/commits/18b0eb44af82d5925c371d30ff75af6c288a0069

- Added necessary APIs to do time generation, manipulations, and conversions.

Steps for migration from the previous version to the current version are listed [here](https://github.com/ballerina-platform/ballerina-standard-library/issues/1079).

#### Cache Package Updates

> - Introduced the new configuration as `EvictionPolicy` to set the eviction policy in the `CacheConfig` record.

Reviewer: ```suggestion
- Introduced the new `EvictionPolicy` configuration to set the eviction policy in the `CacheConfig` record.
```

@praneesha thoughts please?

The `EvictionPolicy` record has been introduced with the option `LRU` as the module only supports the LRU eviction policy to evict the cache data when the cache is full.

> - Removed the `AbstractEvictionPolicy` object.

Reviewer: Was this an object type or a class? Shall we use the appropriate term here?

Same for the description below.

Reviewer 3: This was an object type. Please refer [this](https://github.com/ballerina-platform/module-ballerina-cache/blob/release-2.0.2/cache-ballerina/abstract_eviction_policy.bal#L19).

Reviewer: Then shall we say
```suggestion
- Removed the `AbstractEvictionPolicy` object type.
```

> This object had the common APIs for the cache eviction functionalities to implement the custom eviction policy. It has been removed by introducing the above configuration.

Reviewer: ```suggestion
This object had the common APIs for the cache eviction functionalities to implement the custom eviction policy. It has been removed with the introduction of the above configuration.
```

#### New `xmldata` Package

> A new module is added to convert a natural representation of data in XML into a natural representation of data in JSON and vice-versa.

Reviewer: Shouldn't this be something like
```suggestion
A new module is added to convert a natural representation of data in XML format into a natural representation of data in JSON format and vice-versa.
```

or just
```suggestion
A new module is added to convert data in XML format to JSON format and vice-versa.
```

- Converts a JSON object to an XML representation.
```ballerina
json data = {
    name: "John",
    age: 30
};

> xml|Error x = fromJson(data);

Reviewer: Needs to be a qualified identifier, right? With the import? Please check the other example too.

    second: 50.52,
    timeAbbrev: "Asia/Colombo",
    utcOffset: zoneOffset
};
task:Error|task:JobId id = task:scheduleOneTimeJob(new MyJob(), time);
```

- Schedules the recurring Ballerina job according to the given duration.
```ballerina
import ballerina/task;

class MyJob {
   *task:Job;

   public function execute() {
       // logic goes here
   }
}
task:Error|task:JobId id = task:scheduleJobRecurByFrequency(new MyJob(), 1);
```

- Unschedules the particular job.
```ballerina
import ballerina/task;

task:Error? result = task:unscheduleJob(id);
```

- Pauses all the jobs.
```ballerina
import ballerina/task;

task:Error? result = task:pauseAllJobs();
```

- Resumes all the jobs.
```ballerina
import ballerina/task;

task:Error? result = task:resumeAllJobs();
```

- Pauses the particular job.
```ballerina
import ballerina/task;

task:Error? result = task:pauseJob(id);
```

- Resumes the particular job.
```ballerina
import ballerina/task;

task:Error? result = task:resumeJob(id);
```

- Gets all the running jobs.
```ballerina
import ballerina/task;

task:Error? result = task:getRunningJobs();
```

#### Time Package Updates

Revamped the entire time package as follows:

> - Introduced the `time:Utc` record to represent the UTC timestamp.
> - Introduced the `time:Civil` record to represent the localized time.

Reviewer 4: ```suggestion
- Introduced the `time:Utc` record to represent the UTC timestamp.
- Introduced the `time:Civil` record to represent the localized time.
```

#### Standard Library

##### Log Package Updates

###### Introduce additional log levels and log functions

There are 4 log levels: `DEBUG`, `ERROR`, `INFO`, and `WARN` and their respective functions: `printDebug`, `printError`, `printInfo`, and `printWarn`. To set the global log level, place the entry given below in the `Config.toml` file:

> ```

Reviewer: ```suggestion
```toml
```

[log] 
level = "[LOG_LEVEL]"
```

Each module can also be assigned its own log level. To assign a log level to a module, provide the following entry in the `Config.toml` file:

> ```

Reviewer: ```suggestion
```toml
```

```ballerina
import ballerina/task;

class MyJob {
   *task:Job;

   public function execute() {
       // logic goes here
   }
}
task:Error|task:JobId id = task:scheduleJobRecurByFrequency(new MyJob(), 1);
```

- Unschedules the particular job.
```ballerina
import ballerina/task;

task:Error? result = task:unscheduleJob(id);
```

- Pauses all the jobs.
```ballerina
import ballerina/task;

task:Error? result = task:pauseAllJobs();
```

- Resumes all the jobs.
```ballerina
import ballerina/task;

task:Error? result = task:resumeAllJobs();
```

- Pauses the particular job.
```ballerina
import ballerina/task;

task:Error? result = task:pauseJob(id);
```

- Resumes the particular job.
```ballerina
import ballerina/task;

task:Error? result = task:resumeJob(id);
```

- Gets all the running jobs.
```ballerina
import ballerina/task;

task:Error? result = task:getRunningJobs();
```

#### Time Package Updates

Revamped the entire time package as follows:

- Introduced the `time:Utc` record to represent the UTC timestamp.
- Introduced the `time:Civil` record to represent the localized time.
- Added necessary APIs to do time generation, manipulations, and conversions.

> Steps for migration from the previous version to the current version are listed [here](https://github.com/ballerina-platform/ballerina-standard-library/issues/1079).

Author: ```suggestion
Steps for migration from the previous version to the current version are listed [in this issue](https://github.com/ballerina-platform/ballerina-standard-library/issues/1079).
```

#### Cache Package Updates

- Introduced the new `EvictionPolicy` configuration to set the eviction policy in the `CacheConfig` record.

The `EvictionPolicy` record has been introduced with the option `LRU` as the module only supports the LRU eviction policy to evict the cache data when the cache is full.

- Removed the `AbstractEvictionPolicy` object.

> This object had the common APIs for the cache eviction functionalities to implement the custom eviction policy. It has been removed with the introduction of the above configuration.

Reviewer: ```suggestion
This object type had the common APIs for the cache eviction functionalities to implement a custom eviction policy. It has been removed with the introduction of the above configuration.
```

> ###### All the timeout configurations are converted to accept decimal values and the time unit is in seconds.

Reviewer 2: ```suggestion
###### All the timeout configurations are converted to accept decimal values and the time unit is in seconds.
```

Complete sentences shouldn't be added as titles.

Reviewer 5: I fixed it with https://github.com/ballerina-platform/ballerina-release/pull/115
